text/0000-on-modifier.md

```diff
@@ -0,0 +1,193 @@
+
+```hbs
+<div {{on "click" this.handleClick passive=true}}></div>
+```
+
+```js
+element.addEventListener('click', this.handleClick, { passive: true });
+```
+
+Multiple event listeners can be added to an element by adding the modifier more
+than once:
+
+```hbs
+<div
+  {{on "click" this.handleClick}}
+  {{on "mouseenter" this.handleMouseEnter}}
+>
+</div>
+```
+
+Extra positional parameters will be ignored. In order to pass values to a
+function, users should use a helper such as the [bind helper](https://github.com/emberjs/rfcs/pull/470):
+
+```hbs
+<div {{on "click" (bind this.addNumber 123)}}></div>
+```
+
+`{{on}}` does _not_ bind context. Context binding must be handled via a helper,
+or in the component definition via the `@action` decorator, which will be the
+recommended path:
+
+```js
+class ClickComponent extends Component {
+  @action
+  handleClick() {
+    // ...
+  }
+}
+```
+
+## How we teach this
+
+`{{on}}` itself can be taught in conjuction with event listeners. We can give
+users an overview of event listeners and the `addEventListener` API, and then
+relate the modifier directly to it. We should provide lots of examples of usage
+throughout the guides for how users can use it, including some examples of how
+to apply multiple event listeners to an element.
+
+`{{on}}` also does not bind context, and we should make it clear that the
+`@action` decorator is the recommended way to bind methods to their components.
+
+Other than that, most examples in the guides can be translated directly from
+the `{{action}}` modifier to the `{{on}}` modifier.
+
+## Drawbacks
+
+- This API is somewhat more verbose than the `{{action}}` modifier since it
+  requires a helper for passing additional values to the function. However, this
+  explicitness makes it much more clear what the separation of concerns is, and
+  allows users to fine tune their event handling.
+- The `{{on}}` modifier requires another helper to pass values. Currently, the
+  ideal companion helper is the `{{bind}}` helper, which is still in RFC.
+- The `onclick=` style at first looks like a native browser API, and is
+  sometimes easier to teach because of this. However, it is _not_ the same as
+  the `onclick` attribute, which only receives strings, and the differences are
+  often confusing. In addition, the fact that there are many use cases that
+  aren't covered by these properties, and the fact that they are hostile to SSR,
+  makes them less than ideal.
+
+## Alternatives
+
+One option would be to allow the `{{on}}` modifier to receive event listeners
+as named parameters instead:
+
+```hbs
+<div
+  {{on
+    click=this.handleClick
+    mouseenter=this.handleMouseEnter
+  }}
+>
+</div>
+```
+
+One downside of this style of API is that there is no clear way to pass
+additional options to the individual event listeners. We could pass them as a
+positional parameter, but they would apply to _all_ listeners, which would
+require multiple invocations.
```

Comment:
Couldn't this be solved with `{{on click=(bind this.handleClick 1 2 3)}}`, as is proposed elsewhere in the rfc also?

Reply:
I think @pzuraq means extra parameters like `pasive`, `capture`, etc.

Reply:
Yep, exactly. I don't think it would be possible to pass two different sets of options. You would have to add two `{{on}}` modifiers with both designs:

```hbs
<div
  {{on "click" this.handleClick passive=true}}
  {{on "mouseenter" this.handleMouseEnter passive=false}}
>
</div>

<div
  {{on click=this.handleClick options=(hash passive=true)}}
  {{on mouseenter=this.handleMouseEnter options=(hash passive=false)}}
>
</div>
```

Personally, I think the first form is preferable here, I find the passing of options is clearer (you could imagine someone being confused, thinking there is an `options` event), and it's not _that_ much more verbose when you want to add multiple event listeners.

Reply:
Agreed. separate modifers for separate things is better. I imagine that will enable some syntax sugar as well.